### PR TITLE
Actor timer/reminder responses: marshal data to JSON

### LINKED ISSUE #6268: Actor Timers/Reminders not being unmarshaled correctly
## In what area(s)?

/area runtime

## What version of Dapr?

master 

## Expected Behavior

Reminders and Timers are triggered as expected. 

## Actual Behavior

```
time="2023-04-20T03:39:07.462656618Z" level=warning msg="Redis does not support transaction rollbacks and should not be used in production as an actor state store." app_id=actorreminderfailoveritone-myactorservice component="statestore (state.redis/v1)" instance=fv-az563-898 scope=dapr.contrib type=log ver=edge
== APP == 03:39:08.020 [http-nio-41345-exec-7] ERROR o.a.c.c.C.[.[.[.[dispatcherServlet] - Servlet.service() for servlet [dispatcherServlet] threw exception
== APP == com.fasterxml.jackson.databind.exc.InvalidFormatException: Cannot access contents of TextNode as binary due to broken Base64 encoding: Unexpected end of base64-encoded String: base64 variant 'MIME-NO-LINEFEEDS' expects padding (one or more '=' characters) at the end
== APP == 	at com.fasterxml.jackson.databind.exc.InvalidFormatException.from(InvalidFormatException.java:67)
== APP == 	at com.fasterxml.jackson.databind.node.TextNode.getBinaryValue(TextNode.java:72)
== APP == 	at com.fasterxml.jackson.databind.node.TextNode.binaryValue(TextNode.java:83)
== APP == 	at io.dapr.actors.runtime.ActorObjectSerializer.deserializeActorReminder(ActorObjectSerializer.java:211)
== APP == 	at io.dapr.actors.runtime.ActorObjectSerializer.deserialize(ActorObjectSerializer.java:168)
== APP == 	at io.dapr.actors.runtime.ActorManager.lambda$invokeReminder$4(ActorManager.java:104)
== APP == 	at reactor.core.publisher.MonoSupplier.call(MonoSupplier.java:67)
== APP == 	at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap(FluxFlatMap.java:127)
== APP == 	at reactor.core.publisher.MonoFlatMap.subscribeOrReturn(MonoFlatMap.java:53)
== APP == 	at reactor.core.publisher.Mono.subscribe(Mono.java:4429)
== APP == 	at reactor.core.publisher.MonoIgnorePublisher.subscribe(MonoIgnorePublisher.java:57)
== APP == 	at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:165)
== APP == 	at reactor.core.publisher.FluxHide$SuppressFuseableSubscriber.onNext(FluxHide.java:137)
== APP == 	at reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.complete(MonoIgnoreThen.java:292)
== APP == 	at reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.onNext(MonoIgnoreThen.java:187)
== APP == 	at reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.subscribeNext(MonoIgnoreThen.java:236)
== APP == 	at reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.onComplete(MonoIgnoreThen.java:203)
== APP == 	at reactor.core.publisher.Operators.complete(Operators.java:137)
== APP == 	at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap(FluxFlatMap.java:145)
== APP == 	at reactor.core.publisher.MonoFlatMap.subscribeOrReturn(MonoFlatMap.java:53)
== APP == 	at reactor.core.publisher.Mono.subscribe(Mono.java:4429)
== APP == 	at reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.subscribeNext(MonoIgnoreThen.java:263)
== APP == 	at reactor.core.publisher.MonoIgnoreThen.subscribe(MonoIgnoreThen.java:51)
== APP == 	at reactor.core.publisher.Mono.subscribe(Mono.java:4444)
== APP == 	at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap(FluxFlatMap.java:203)
== APP == 	at reactor.core.publisher.MonoFlatMap.subscribeOrReturn(MonoFlatMap.java:53)
== APP == 	at reactor.core.publisher.Mono.subscribe(Mono.java:4429)
== APP == 	at org.springframework.web.servlet.mvc.method.annotation.ReactiveTypeHandler$DeferredResultSubscriber.connect(ReactiveTypeHandler.java:442)
== APP == 	at org.springframework.web.servlet.mvc.method.annotation.ReactiveTypeHandler.handleValue(ReactiveTypeHandler.java:157)
== APP == 	at org.springframework.web.servlet.mvc.method.annotation.ResponseBodyEmitterReturnValueHandler.handleReturnValue(ResponseBodyEmitterReturnValueHandler.java:159)
== APP == 	at org.springframework.web.method.support.HandlerMethodReturnValueHandlerComposite.handleReturnValue(HandlerMethodReturnValueHandlerComposite.java:82)
== APP == 	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:123)
== APP == 	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:878)
== APP == 	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:792)
== APP == 	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)
== APP == 	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1040)
== APP == 	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:943)
== APP == 	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
== APP == 	at org.springframework.web.servlet.FrameworkServlet.doPut(FrameworkServlet.java:920)
== APP == 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:655)
== APP == 	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
== APP == 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:733)
== APP == 	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
== APP == 	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
== APP == 	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
== APP == 	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
== APP == 	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
== APP == 	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100)
== APP == 	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
== APP == 	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
== APP == 	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
== APP == 	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93)
== APP == 	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
== APP == 	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
== APP == 	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
== APP == 	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
== APP == 	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
== APP == 	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
== APP == 	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
== APP == 	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:202)
== APP == 	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97)
== APP == 	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:542)
== APP == 	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:143)
== APP == 	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
== APP == 	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78)
== APP == 	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
== APP == 	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:374)
== APP == 	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
== APP == 	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:868)
== APP == 	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1590)
== APP == 	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
== APP == 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
== APP == 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
== APP == 	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
== APP == 	at java.base/java.lang.Thread.run(Thread.java:829)
```

Related Java IT run where we get the same exception 
https://github.com/dapr/java-sdk/actions/runs/4749652532/jobs/8438185954?pr=846#step:19:3009

## Steps to Reproduce the Problem

Run Java ITs against latest commit of Dapr.

## Release Note

RELEASE NOTE:

### PR DESCRIPTION
**Pull request by ItalyPaleAle**
Fixes #6268
Fixes dapr/go-sdk#390

### CODE REVIEW

**Comment by mukundansundar** [comment]
I have checked with Java SDK, this seems to be working as expected 
```
INFO[0007] placement tables updated, version: 0          app_id=demoactorservice instance=localhost scope=dapr.runtime.actor.internal.placement type=log ver=edge
INFO[0009] placement tables updated, version: 1          app_id=demoactorservice instance=localhost scope=dapr.runtime.actor.internal.placement type=log ver=edge
== APP == 2023-04-21 15:36:32,046 {HH:mm:ss.SSS} [http-nio-3000-exec-5] INFO  io.dapr.actors.ActorTrace - Actor:81ed5338-107f-46f3-a2c2-bac50f041aa9 Activating ...
== APP == 2023-04-21 15:36:32,046 {HH:mm:ss.SSS} [http-nio-3000-exec-6] INFO  io.dapr.actors.ActorTrace - Actor:c957d860-12d7-4079-bb68-3d73ddb6a267 Activating ...
== APP == 2023-04-21 15:36:32,046 {HH:mm:ss.SSS} [http-nio-3000-exec-7] INFO  io.dapr.actors.ActorTrace - Actor:6fb472cc-287d-4554-b13e-a1d4cb8cca5b Activating ...
== APP == 2023-04-21 15:36:32,046 {HH:mm:ss.SSS} [http-nio-3000-exec-5] INFO  io.dapr.actors.ActorTrace - Actor:81ed5338-107f-46f3-a2c2-bac50f041aa9 Activated
== APP == 2023-04-21 15:36:32,046 {HH:mm:ss.SSS} [http-nio-3000-exec-6] INFO  io.dapr.actors.ActorTrace - Actor:c957d860-12d7-4079-bb68-3d73ddb6a267 Activated
== APP == 2023-04-21 15:36:32,046 {HH:mm:ss.SSS} [http-nio-3000-exec-7] INFO  io.dapr.actors.ActorTrace - Actor:6fb472cc-287d-4554-b13e-a1d4cb8cca5b Activated
WARN[0023] Redis does not support transaction rollbacks and should not be used in production as an actor state store.  app_id=demoactorservice component="statestore (state.redis/v1)" instance=localhost scope=dapr.contrib type=log ver=edge
== APP == Server say method for actor 6fb472cc-287d-4554-b13e-a1d4cb8cca5b: Actor 6fb472cc-287d-4554-b13e-a1d4cb8cca5b said message #1 @ 2023-04-21 15:36:32.146
== APP == Server say method for actor c957d860-12d7-4079-bb68-3d73ddb6a267: Actor c957d860-12d7-4079-bb68-3d73ddb6a267 said message #1 @ 2023-04-21 15:36:32.146
== APP == Server say method for actor 81ed5338-107f-46f3-a2c2-bac50f041aa9: Actor 81ed5338-107f-46f3-a2c2-bac50f041aa9 said message #1 @ 2023-04-21 15:36:32.146
== APP == Server say method for actor c957d860-12d7-4079-bb68-3d73ddb6a267: Actor c957d860-12d7-4079-bb68-3d73ddb6a267 said message #2 @ 2023-04-21 15:36:32.411
== APP == Server say method for actor 81ed5338-107f-46f3-a2c2-bac50f041aa9: Actor 81ed5338-107f-46f3-a2c2-bac50f041aa9 said message #2 @ 2023-04-21 15:36:32.460
== APP == Server say method for actor 81ed5338-107f-46f3-a2c2-bac50f041aa9: Actor 81ed5338-107f-46f3-a2c2-bac50f041aa9 said message #3 @ 2023-04-21 15:36:32.489
== APP == Server say method for actor c957d860-12d7-4079-bb68-3d73ddb6a267: Actor c957d860-12d7-4079-bb68-3d73ddb6a267 said message #3 @ 2023-04-21 15:36:32.681
== APP == Server say method for actor 6fb472cc-287d-4554-b13e-a1d4cb8cca5b: Actor 6fb472cc-287d-4554-b13e-a1d4cb8cca5b said message #2 @ 2023-04-21 15:36:32.723
== APP == Server say method for actor c957d860-12d7-4079-bb68-3d73ddb6a267: Actor c957d860-12d7-4079-bb68-3d73ddb6a267 said message #4 @ 2023-04-21 15:36:33.079
== APP == Server say method for actor 81ed5338-107f-46f3-a2c2-bac50f041aa9: Actor 81ed5338-107f-46f3-a2c2-bac50f041aa9 said message #4 @ 2023-04-21 15:36:33.169
== APP == Server say method for actor 6fb472cc-287d-4554-b13e-a1d4cb8cca5b: Actor 6fb472cc-287d-4554-b13e-a1d4cb8cca5b said message #3 @ 2023-04-21 15:36:33.332
== APP == Server say method for actor 81ed5338-107f-46f3-a2c2-bac50f041aa9: Actor 81ed5338-107f-46f3-a2c2-bac50f041aa9 said message #5 @ 2023-04-21 15:36:33.770
== APP == Server say method for actor 6fb472cc-287d-4554-b13e-a1d4cb8cca5b: Actor 6fb472cc-287d-4554-b13e-a1d4cb8cca5b said message #4 @ 2023-04-21 15:36:33.851
== APP == Server say method for actor c957d860-12d7-4079-bb68-3d73ddb6a267: Actor c957d860-12d7-4079-bb68-3d73ddb6a267 said message #5 @ 2023-04-21 15:36:33.957
== APP == Server timer for actor 81ed5338-107f-46f3-a2c2-bac50f041aa9: ping! @ 2023-04-21 15:36:34.010
== APP == Server timer for actor c957d860-12d7-4079-bb68-3d73ddb6a267: ping! @ 2023-04-21 15:36:34.010
== APP == Server timer for actor 6fb472cc-287d-4554-b13e-a1d4cb8cca5b: ping! @ 2023-04-21 15:36:34.010
== APP == Server say method for actor 6fb472cc-287d-4554-b13e-a1d4cb8cca5b: Actor 6fb472cc-287d-4554-b13e-a1d4cb8cca5b said message #5 @ 2023-04-21 15:36:34.283
== APP == Server say method for actor 6fb472cc-287d-4554-b13e-a1d4cb8cca5b: Actor 6fb472cc-287d-4554-b13e-a1d4cb8cca5b said message #6 @ 2023-04-21 15:36:34.360
== APP == Server say method for actor 81ed5338-107f-46f3-a2c2-bac50f041aa9: Actor 81ed5338-107f-46f3-a2c2-bac50f041aa9 said message #6 @ 2023-04-21 15:36:34.766
== APP == Server say method for actor c957d860-12d7-4079-bb68-3d73ddb6a267: Actor c957d860-12d7-4079-bb68-3d73ddb6a267 said message #6 @ 2023-04-21 15:36:34.844
== APP == Server timer for actor 81ed5338-107f-46f3-a2c2-bac50f041aa9: ping! @ 2023-04-21 15:36:35.001
== APP == Server timer for actor c957d860-12d7-4079-bb68-3d73ddb6a267: ping! @ 2023-04-21 15:36:35.001
```

File: pkg/actors/requests_responses.go [hunk 1]
```diff
@@ -104,3 +124,21 @@ type TimerResponse struct {
 	DueTime  string `json:"dueTime"`
 	Period   string `json:"period"`
 }
+
+// MarshalJSON is a custom JSON marshaler that encodes the data as JSON.
+func (t *TimerResponse) MarshalJSON() ([]byte, error) {
```
Comment (mukundansundar) [inline pkg/actors/requests_responses.go:129]: Please add a comment against any MarshalJSON() function with this signature that the Marshaller interface used by `encoding/json` is being implemented with this function.

File: pkg/actors/requests_responses.go [hunk 2]
```diff
@@ -76,6 +78,24 @@ type ReminderResponse struct {
...
+	}{
+		responseAlias: (*responseAlias)(r),
+	}
+
+	var err error
+	m.Data, err = json.Marshal(r.Data)
+	if err != nil {
+		return nil, err
+	}
+	return json.Marshal(m)
```
Comment (mukundansundar) [inline pkg/actors/requests_responses.go:96]: why is a double JSON marshal needed?

File: pkg/actors/requests_responses.go [hunk 3]
```diff
@@ -76,6 +78,24 @@ type ReminderResponse struct {
 	Period  string `json:"period"`
 }
 
+// MarshalJSON is a custom JSON marshaler that encodes the data as JSON.
+func (r *ReminderResponse) MarshalJSON() ([]byte, error) {
+	type responseAlias ReminderResponse
+	m := struct {
+		Data []byte `json:"data,omitempty"`
+		*responseAlias
```
Comment (mukundansundar) [inline pkg/actors/requests_responses.go:86]: could you explain the need for this local type definition and the embedding here?

**Comment by ItalyPaleAle** [comment]
Explanation of the issue.

Actor SDKs send reminders and timers data in a way that is JSON-encoded. From the Go SDK example:
image
From the .NET SDK example:
image

Actor SDKs then send the data via gRPC to Dapr, and the data is sent in a field of type `[]bytes`. No actor SDK uses HTTP, but I verified by invoking the HTTP APIs that it expects the same thing: the `data` property is expected to be a JSON-encoded string (although this is undocumented, all actor SDKs assume that).

In 1.10.4 the field was then stored in the database using a struct where the data was `[]byte`. In 1.10.5 that changed to `json.RawMessage`. The representation in the database is the same, since `json.RawMessage` is `[]byte`.

However when the data is sent to the actor host using HTTP, it's serialized as JSON. That's where `json.RawMessage` is serialized as the string itself, while `[]byte` is serialized as a base64-encoded JSON representation.

```json
// 1.10.4
{"data":"eyJoZWxsbyI6IndvcmxkIn0="}

// 1.10.5
{"data":{"hello":"world"}}
```

Although the second is much nicer (and efficient), it was unexpected by the SDKs. Although this was not in the contract (and probably not intended either), all SDKs seem to depend on this.

The fix is to make sure that the data is returned to the clients in a `[]byte` so when it's serialized as JSON it becomes base64-encoded.

File: pkg/actors/requests_responses.go (hunk 1 above)
Comment (ItalyPaleAle) [reply pkg/actors/requests_responses.go:129]: Done

File: pkg/actors/requests_responses.go (hunk 3 above)
Comment (ItalyPaleAle) [reply pkg/actors/requests_responses.go:86]: This is how you implement a JSON marshaller that overrides a single field. We do this in other parts around Dapr and it's a common pattern in Go. Example: http://choly.ca/post/go-json-marshalling/

File: pkg/actors/requests_responses.go (hunk 2 above)
Comment (ItalyPaleAle) [reply pkg/actors/requests_responses.go]: Explained below and in the comments on the code.

File: pkg/actors/internal_actor.go [hunk 4]
```diff
@@ -112,31 +112,33 @@ func (c *internalActorChannel) InvokeMethod(ctx context.Context, req *invokev1.I
...
-			return nil, err
-		}
-
 		// Check for well-known method names; otherwise, just call InvokeMethod on the internal actor.
 		if strings.HasPrefix(methodName, "remind/") {
 			reminderName := strings.TrimPrefix(methodName, "remind/")
-			reminderInfo := new(ReminderResponse)
-			if err = json.Unmarshal(requestData, reminderInfo); err != nil {
-				return nil, fmt.Errorf("unexpected ReminderResponse JSON payload: %s", requestData)
+			reminderInfo, ok := req.GetDataObject().(*ReminderResponse)
```
Comment (ItalyPaleAle) [inline pkg/actors/internal_actor.go:118]: This is needed to make tests pass.

Also avoids the unnecessary step of JSON marshalling and unmarshalling

**Comment by artursouza** [comment]
Is this a hotfix?

**Comment by berndverst** [comment]
@ItalyPaleAle the Python Actor SDK is HTTP only. And I thought that was HTTP also in DotNet?

**Comment by berndverst** [comment]
FYI: I manually built the Dapr sidecar from this PR and ran the Python SDK actor integration test. Still passes.

**Comment by ItalyPaleAle** [comment]
Thanks for confirming. Although the SDK is HTTP only, looking at the comments it still needs the body to be a JSON-encoded string:

image

So that would still work the same way here.

**Review by artursouza — CHANGES_REQUESTED** [review]
Merging into 1.10 release first.

**Comment by artursouza** [comment]
Closing in favor of https://github.com/dapr/dapr/pull/6284/files

**Pull request closed without merging**